bot/utils/redis_cache.py

```diff
@@ -0,0 +1,360 @@
+        return await self._redis.hdel(self._namespace, key)
+
+    async def contains(self, key: RedisType) -> bool:
+        """
+        Check if a key exists in the Redis cache.
+
+        Return True if the key exists, otherwise False.
+        """
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        exists = await self._redis.hexists(self._namespace, key)
+
+        log.trace(f"Testing if {key} exists in the RedisCache - Result is {exists}")
+        return exists
+
+    async def items(self) -> ItemsView:
+        """
+        Fetch all the key/value pairs in the cache.
+
+        Returns a normal ItemsView, like you would get from dict.items().
+
+        Keep in mind that these items are just a _copy_ of the data in the
+        RedisCache - any changes you make to them will not be reflected
+        into the RedisCache itself. If you want to change these, you need
+        to make a .set call.
+
+        Example:
+        items = await my_cache.items()
+        for key, value in items:
+            # Iterate like a normal dictionary
+        """
+        await self._validate_cache()
+        items = self._dict_from_typestring(
+            await self._redis.hgetall(self._namespace)
+        ).items()
+
+        log.trace(f"Retrieving all key/value pairs from cache, total of {len(items)} items.")
+        return items
+
+    async def length(self) -> int:
+        """Return the number of items in the Redis cache."""
+        await self._validate_cache()
+        number_of_items = await self._redis.hlen(self._namespace)
+        log.trace(f"Returning length. Result is {number_of_items}.")
+        return number_of_items
+
+    async def to_dict(self) -> Dict:
+        """Convert to dict and return."""
+        return {key: value for key, value in await self.items()}
+
+    async def clear(self) -> None:
+        """Deletes the entire hash from the Redis cache."""
+        await self._validate_cache()
+        log.trace("Clearing the cache of all key/value pairs.")
+        await self._redis.delete(self._namespace)
+
+    async def pop(self, key: RedisType, default: Optional[RedisType] = None) -> RedisType:
+        """Get the item, remove it from the cache, and provide a default if not found."""
+        log.trace(f"Attempting to pop {key}.")
+        value = await self.get(key, default)
+
+        # No need to try to delete something that doesn't exist,
+        # that's just a superfluous API call.
+        if value != default:
+            log.trace(f"Key {key} exists, deleting it from the cache.")
+            await self.delete(key)
+
+        return value
+
+    async def update(self, items: Dict) -> None:
+        """Update the Redis cache with multiple values."""
+        await self._validate_cache()
+        log.trace(f"Updating the cache with the following items:\n{items}")
+        await self._redis.hmset_dict(self._namespace, self._dict_to_typestring(items))
```

Comment:
If I understand this correctly, the `items` param could be annotated as `Dict[RedisType, ReditType]`.

Reply:
I'm aware that this mimics a dict update, but maybe the docstring could be a little bit more detailed? If you think it's fine as it is, then it can stay.

I didn't instantly understand what this will do from the function signature & docstring, but I also didn't connect it with a dict at first.